src/framework/smoke_queue.hpp

```diff
@@ -0,0 +1,171 @@
+#ifndef SERENITY_SMOKE_QUEUE_HPP
+#define SERENITY_SMOKE_QUEUE_HPP
+
+#include <memory>
+#include <string>
+#include <random>
+#include <vector>
+
+#include <mesos/resources.hpp>
+
+#include <stout/json.hpp>
+#include <stout/option.hpp>
+
+#include "logging/logging.hpp"
+
+#include "smoke_job.hpp"
+
+struct QueueBlock {
+  QueueBlock(std::shared_ptr<SmokeJob> _job,
+              double_t _probability,
+              size_t _index)
+    : baseJob(_job),
+      originalProbability(_probability),
+      baseProbability(_probability),
+      index(_index),
+      aliasJob(nullptr),
+      ready(false) {}
+
+  double_t originalProbability;
+  std::shared_ptr<SmokeJob> baseJob;
+  double_t baseProbability;
+
+  std::shared_ptr<SmokeJob> aliasJob;
+
+  size_t index;
+  bool ready;
+
+  std::string toString() const {
+    std::stringstream result;
+    result << "{ id: " << baseJob->id
+    << ", baseProbability: " << baseProbability
+    << ", aliasId: "
+    << (aliasJob != nullptr ? std::to_string(aliasJob->id) : "<none>")
+    << "}";
+    return result.str();
+  }
+
+};
+
+
+class SmokeAliasQueue {
+ public:
+  SmokeAliasQueue() : totalShares(0), initialized(false),
+                 localIndex(0), finished(false) {}
+
+  /**
+   * Using Naive Alias Method from http://www.keithschwarz.com/darts-dice-coins/
+   */
+  void init() {
```

Comment:
If init is not needed by others, then you should protect it.
Also, it could be called initAliasParameters.